### PR TITLE
NMS-13016: run as non-root by default

### PR DESCRIPTION
**Pull request by RangerRick**
This PR changes OpenNMS to run as `opennms` by default in both the debian and RPM packages.

It does a number of things:

* adds some command-line shell utilities for fixing/changing permissions
* sets `$RUNAS` in all scripts to `opennms`
* updates the `ensure-user-ping.sh` script to handle updating the group to match the `opennms` GID
* adds logic to `$OPENNMS_HOME/bin/install` to validate ownership

By default, it _only_ changes the permissions of files _installed_ by the packages, it does _not_ recurse through the user's data, since large installs could take hours to complete a recursive chown.

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13016

### CODE REVIEW

**Comment by agalue** [comment]
I'll try to find time to test the DEB and RPM packages on a fresh install and through an upgrade to see how all this works.

However, I cannot see the changes to the `Dockerfile` for OpenNMS now that we don't need to apply the changes there:

https://github.com/OpenNMS/opennms/blob/develop/opennms-container/horizon/Dockerfile#L38-L39
https://github.com/OpenNMS/opennms/blob/develop/opennms-container/horizon/Dockerfile#L116-L120
https://github.com/OpenNMS/opennms/blob/develop/opennms-container/horizon/Dockerfile#L133-L134

The following line should be consistent with the package manager:

https://github.com/OpenNMS/opennms/blob/develop/opennms-container/horizon/Dockerfile#L161

**Comment by j-white** [comment]
What is the expected behavior for `trapd` (binds to port 162) and for ICMP pinging now that the privileges are dropped?

**Comment by RangerRick** [comment]
> I'll try to find time to test the DEB and RPM packages on a fresh install and through an upgrade to see how all this works.
> 
> However, I cannot see the changes to the `Dockerfile` for OpenNMS now that we don't need to apply the changes there:
> 
> https://github.com/OpenNMS/opennms/blob/develop/opennms-container/horizon/Dockerfile#L38-L39
> https://github.com/OpenNMS/opennms/blob/develop/opennms-container/horizon/Dockerfile#L116-L120
> https://github.com/OpenNMS/opennms/blob/develop/opennms-container/horizon/Dockerfile#L133-L134
> 
> The following line should be consistent with the package manager:
> 
> https://github.com/OpenNMS/opennms/blob/develop/opennms-container/horizon/Dockerfile#L161

Good point, I hadn't looked into the docker images, I thought they were already running as non-root like minion/sentinel.  I'll take a look.

> What is the expected behavior for `trapd` (binds to port 162) and for ICMP pinging now that the privileges are dropped?

ICMP should be handled by the `ensure-user-ping.sh` script, which is called in the package postinstall.  Trapd I guess we'll have to set higher and then add some documentation on port-forwarding?

**Review by indigo423 — COMMENTED** [review]
The only thing is 162/udp for trapd as @j-white mentioned as well. In docker, we use 1162 similar to 1514/udp for Syslogd. In containers, Docker takes care of redirects. Suggestion to solve this, change from 162/udp to 1162/udp and explain in our docs, based on here: https://opennms.discourse.group/t/how-to-deal-with-privileged-ports-1024-for-network-management-functions/1572.

Ping we can follow the procedure here, especially if we set UID/GID: https://opennms.discourse.group/t/how-to-allow-unprivileged-users-to-use-icmp-ping/1573.

If we try to cover it in the docs, I would prefer the options **without** giving capabilities to the `java` binary with `set_cap`. We get a wider attack vector through any program running with `java` and can mention it as the last option with a bigger warning sign if don't have any other option.

File: debian/opennms-common.postinst [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#! /bin/sh
+# postinst script for opennms-common
+#
+# vim:set ai et sts=4 sw=4 tw=0:
+
+## create the OpenNMS group and user if it doesn't exist
+getent group opennms >/dev/null 2>&1 || addgroup --system opennms
+getent passwd opennms >/dev/null 2>&1 ||
+    adduser --system --home /usr/share/opennms --no-create-home \
+            --disabled-password --ingroup opennms opennms
```
[the remark this answers is not in the document]
Comment (RangerRick) [reply debian/opennms-common.postinst:10]: On a regular system we can't really assume they'll be available.  We could try _preferring_ it but I don't think we could rely on assuming it matches.

**Comment by dino2gnt** [comment]
There is a process provided by the various distros for application packagers to request allocation of a known UID/GID combo.  Do we care about that, or are we content to dynamically allocate a UID and GID from the user range instead? This could lead to inconsistent UIDs across installs.  Does that matter? 

Ex: https://docs.fedoraproject.org/en-US/packaging-guidelines/UsersAndGroups/#Soft_static_allocation

**Comment by RangerRick** [comment]
> There is a process provided by the various distros for application packagers to request allocation of a known UID/GID combo. Do we care about that, or are we content to dynamically allocate a UID and GID from the user range instead? This could lead to inconsistent UIDs across installs. Does that matter?
> 
> Ex: https://docs.fedoraproject.org/en-US/packaging-guidelines/UsersAndGroups/#Soft_static_allocation

I would guess we'd be hard pressed to ask for it considering we're not even submitted as an official package, and we install in `/opt`, but I suppose we could ask?

**Comment by RangerRick** [comment]
OK, the latest updates have changes to attempt to use uid/gid 1000 if available, otherwise fall back to whatever the system gives you. Also ports over some of the perm-fixing from the other packages.

File: opennms-base-assembly/src/main/filtered/etc/trapd-configuration.xml [hunk 2]
```diff
@@ -1 +1 @@
-<trapd-configuration xmlns="http://xmlns.opennms.org/xsd/config/trapd" snmp-trap-address="*" snmp-trap-port="162" new-suspect-on-trap="false" include-raw-message="false" threads="0" queue-size="10000" batch-size="1000" batch-interval="500"/>
\ No newline at end of file
+<trapd-configuration xmlns="http://xmlns.opennms.org/xsd/config/trapd" snmp-trap-address="*" snmp-trap-port="1162" new-suspect-on-trap="false" include-raw-message="false" threads="0" queue-size="10000" batch-size="1000" batch-interval="500"/>
```
Comment (j-white) [inline opennms-base-assembly/src/main/filtered/etc/trapd-configuration.xml:1]: Minion also defaults to 1162 - so these will conflict by default when run on the same host now

File: opennms-install/src/main/java/org/opennms/install/Installer.java [hunk 3]
```diff
@@ -572,6 +577,56 @@ public void verifyFilesAndDirectories() throws FileNotFoundException {
...
+        if (RUNAS_USER == null) {
+            final var opennmsConf = readOpennmsConf();
+            // use RUNAS from opennms.conf if found, fall back to -Dopennms.runas, fall back to "opennms" if nothing is found
+            RUNAS_USER = opennmsConf.getProperty("RUNAS", System.getProperty("opennms.runas", "opennms"));
+        }
+
+        return RUNAS_USER;
+    }
+
+    private Properties readOpennmsConf() throws IOException {
```
Comment (j-white) [inline opennms-install/src/main/java/org/opennms/install/Installer.java:608]: Good candidate for a unit test? Feels like there are a lot of edge cases in here.

File: debian/control [hunk 4]
```diff
@@ -1,13 +1,13 @@
 Source: opennms
 Section: contrib/net
 Priority: optional
-Maintainer: Jeff Gehlbach <jeffg@opennms.org>
+Maintainer: Benjamin Reed <ranger@opennms.org>
```
[the remark this answers is not in the document]
Comment (RangerRick) [reply debian/control:4]: I mean, it's been me for quite a while now  :)

File: .circleci/config.yml [hunk 5]
```diff
@@ -653,40 +653,22 @@ workflows:
             - integration-test
           filters:
             branches:
-              only:
```
[the remark this answers is not in the document]
Comment (RangerRick) [reply .circleci/config.yml:656]: It's probably worth talking about. I'm told we have leftover credits in circle for now, so I thought it might be useful to turn them on all the time. I was gonna take that out when I merged though, it was just to make testing easier/quicker as I was iterating.

File: opennms-install/src/main/java/org/opennms/install/Installer.java (hunk 3 above)
Comment (RangerRick) [reply opennms-install/src/main/java/org/opennms/install/Installer.java:608]: Good point, I'll add one.

File: opennms-base-assembly/src/main/filtered/etc/trapd-configuration.xml (hunk 2 above)
Comment (RangerRick) [reply opennms-base-assembly/src/main/filtered/etc/trapd-configuration.xml:1]: I don't imagine running both on the same system is super common, but doesn't hurt to change it I suppose.

File: opennms-base-assembly/src/main/filtered/etc/trapd-configuration.xml [hunk 6]
```diff
@@ -1 +1 @@
-<trapd-configuration xmlns="http://xmlns.opennms.org/xsd/config/trapd" snmp-trap-address="*" snmp-trap-port="1162" new-suspect-on-trap="false" include-raw-message="false" threads="0" queue-size="10000" batch-size="1000" batch-interval="500"/>
+<trapd-configuration xmlns="http://xmlns.opennms.org/xsd/config/trapd" snmp-trap-address="*" snmp-trap-port="2162" new-suspect-on-trap="false" include-raw-message="false" threads="0" queue-size="10000" batch-size="1000" batch-interval="500"/>
```
Comment (indigo423) [inline opennms-base-assembly/src/main/filtered/etc/trapd-configuration.xml:1]: Just for fun :) What about using 10162 than we use the same pattern as with Syslogd 10514 on core?

https://github.com/OpenNMS/opennms/blob/7ecbfee2a8ea39d250ef3982167de39eee34df99/opennms-base-assembly/src/main/filtered/etc/syslogd-configuration.xml#L9

Comment (RangerRick) [reply opennms-base-assembly/src/main/filtered/etc/trapd-configuration.xml:1]: sure... ¯\_(ツ)_/¯

File: .circleci/config.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (RangerRick) [reply .circleci/config.yml:656]: (Thinking about it, we don't usually need them that often unless it's for specific stuff. I'll revert that change before merging.)

**Comment by indigo423** [comment]
We have to update the docs reflecting the new SNMP Trap UDP listening port. Just a quick sanity check search for 162/udp 
gave me this here:

```
modules/deployment/pages/core/centos-rhel/initialize-core.adoc
57:     By default, the OpenNMS {page-component-title} SNMP trap daemon listens on 162/udp and the Syslog daemon listens on 10514/udp.
```

```
modules/deployment/pages/core/debian-ubuntu/initialize-core.adoc
49:TIP: By default the OpenNMS SNMP Trap daemon listens on 162/udp and the Syslog daemon listens on 10514/udp.
```

**Comment by RangerRick** [comment]
> We have to update the docs reflecting the new SNMP Trap UDP listening port.

Fixed your examples, didn't find anything else obvious with a grep.

I'm inclined to go ahead and merge this unless anyone has other feedback, it's likely there are still things we're going to run into, but I'm not sure we're going to find them without getting these changes out into the wider world.
